File: other/典型90問/001_Yokan Party (★★★★☆☆☆)/a.py
# 001 - Yokan Party (★★★★☆☆☆)

# 最小値の最大値を求める問題は、「2分探索」を行う。

n, l = map(int, input().split())
k = int(input())
a_ = [0] + list(map(int, input().split())) + [l]
a = [a_[i + 1] - a_[i] for i in range(n+1)] 

left = 0
right = l 
mid = (left + right) // 2

while left < right - 1:
    cutnum = 0
    length = 0
    for i in range(n+1):
        length += a[i]
        if length >= mid:
            if cutnum == k:
                continue
            length = 0
            cutnum += 1
    if cutnum < k or length < mid:
        right = mid
    else:
        left = mid
    mid = (left + right) // 2
   
print(mid)

File: other/典型90問/001_Yokan Party (★★★★☆☆☆)/a.cpp
// 001-Yokan Party (★★★★☆☆☆)

// 最小値の最大値を求める問題は、「2分探索」を行う。

# include <bits/stdc++.h>
using namespace std;

int main() {
    int n, l, k;
    cin >> n >> l >> k;
    int _a[n+1], a[n+1];
    _a[0] = 0;
    for (int i = 1; i < n + 1; i++) {
        cin >> _a[i];
        a[i-1] = _a[i] - _a[i-1];
    }
    a[n] = l - _a[n];

    int left = 0, right = l;
    int mid = (left + right) / 2;
    int cutnum, length;

    while (left < right - 1) {
        cutnum = 0;
        length = 0;
        for (int i = 0; i < n + 1; i++) {
            length += a[i];
            if (length >= mid) {
                if (cutnum == k) continue;
                length = 0;
                cutnum += 1;
            }
        }
        if ((cutnum < k) || (length < mid)) right = mid;
        else left = mid;
        mid = (left + right) / 2;
    }
    cout << mid << endl;
    return 0;
}